src/dataflow/src/server/metrics.rs

```diff
@@ -105,36 +105,36 @@ impl WorkerMetrics {
 /// Count of how many metrics we have processed for a given command type
 #[derive(Debug)]
 struct CommandsProcessedMetrics {
-    cache: Vec<i64>,
-    counters: Vec<DeleteOnDropCounter<'static, AtomicI64, Vec<String>>>,
+    worker: String,
+    commands_processed_metric: IntCounterVec,
+    cache: HashMap<&'static str, i64>,
+    counters: HashMap<&'static str, DeleteOnDropCounter<'static, AtomicI64, Vec<String>>>,
 }
 
 impl CommandsProcessedMetrics {
     fn new(worker: &str, commands_processed_metric: &IntCounterVec) -> CommandsProcessedMetrics {
         CommandsProcessedMetrics {
-            cache: CommandKind::into_enum_iter().map(|_| 0).collect(),
-            counters: CommandKind::into_enum_iter()
-                .map(|kind| {
-                    commands_processed_metric.get_delete_on_drop_counter(vec![
-                        worker.to_string(),
-                        kind.name().to_string(),
-                    ])
-                })
-                .collect(),
+            worker: worker.to_string(),
+            commands_processed_metric: commands_processed_metric.clone(),
+            cache: HashMap::new(),
+            counters: HashMap::new(),
```

Comment:
Worth noting that this changes the behavior of metric counters: Previously, it would emit a `0` counter for every kind of command even if no such command had ever been processed. Now, it'll emit a counter only after the first command of its kind.

Reply:
I think we can get this behavior back by adding back that enum_kind derive & front-loading the filling up of command names here; that could also serve as a nice assertion that the command-names-as-metrics are unique.

Reply:
I have weird feels about blurting out zeros for things that don't happen, but that is probably just my streaming background. It would be great if the way that we do this doesn't lock down the types quite as much, as we will probably need to rip them up again soon-ish.